src/flamegraph/attrs.rs

```diff
@@ -0,0 +1,254 @@
+                                            "href" => funcattrs.a.href = Some(value),
+                                            "target" => funcattrs.a.target = Some(value),
+                                            "g_extra" => {
+                                                parse_extra_attrs(&mut funcattrs.g.extra, &value)
+                                            }
+                                            "a_extra" => {
+                                                parse_extra_attrs(&mut funcattrs.a.extra, &value)
+                                            }
+                                            _ => warn!(
+                                                "invalid attribute {} found for {}",
+                                                name, func
+                                            ),
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        Ok(funcattr_map)
+    }
+
+    /// Return FrameAttrs for the given function name if it exists
+    pub fn frameattrs_for_func(&self, func: &str) -> Option<&FrameAttrs> {
+        self.0.get(func)
+    }
+}
+
+/// Attributes to set on the SVG elements of a frame
+#[derive(PartialEq, Eq, Debug, Default)]
+pub struct FrameAttrs {
+    /// The text to include in the `title` element.
+    /// If set to None, the title is dynamically generated based on the function name.
+    pub title: Option<String>,
+
+    pub g: GElementAttrs,
+    pub a: AElementAttrs,
+}
+
+/// Attributes to set on the SVG `g` element.
+/// Any of them set to `None` will get the default value.
+#[derive(PartialEq, Eq, Debug, Default)]
+pub struct GElementAttrs {
+    /// Defaults to "func_g"
+    pub class: Option<String>,
+
+    /// Will not be included if None
+    pub style: Option<String>,
+
+    /// Defaults to "s(this)"
+    pub onmouseover: Option<String>,
+
+    /// Defaults to "c()"
+    pub onmouseout: Option<String>,
+
+    /// Defaults to "zoom(this)"
+    pub onclick: Option<String>,
+
+    /// Extra attributes to include
+    pub extra: HashMap<String, String>,
+}
+
+/// Attributes to set on the SVG `a` element
+#[derive(PartialEq, Eq, Debug, Default)]
+pub struct AElementAttrs {
+    /// If set to None the `a` tag will not be added
+    pub href: Option<String>,
+    pub target: Option<String>,
+
+    /// Extra attributes to include
+    pub extra: HashMap<String, String>,
+}
+
+fn parse_extra_attrs(attrs: &mut HashMap<String, String>, s: &str) {
+    for (name, value) in name_value_pairs(s) {
+        attrs.insert(name, value);
+    }
+}
+
+fn name_value_pairs(s: &str) -> AttrIter {
+    AttrIter { s }
+}
+
+struct AttrIter<'a> {
+    s: &'a str,
+}
+
+impl<'a> Iterator for AttrIter<'a> {
+    type Item = (String, String);
+
+    fn next(&mut self) -> Option<(String, String)> {
```

Comment:
I think you want to take advantage of the fact that you can use `?` with `Option` to make this function nicer to read :)

Reply:
Yes, much better! I'll get a hang of this Rust thing eventually.